cpp/src/strings/json/json_path.cu

```diff
@@ -0,0 +1,957 @@
+          pos++;
+          if (op.name.size_bytes() == 1 && op.name.data()[0] == '*') {
+            op.type          = path_operator_type::CHILD_WILDCARD;
+            op.expected_type = NONE;
+          } else {
+            if (is_string) {
+              op.type          = path_operator_type::CHILD;
+              op.expected_type = OBJECT;
+            } else {
+              op.type  = path_operator_type::CHILD_INDEX;
+              op.index = cudf::io::parse_numeric<int>(
+                op.name.data(), op.name.data() + op.name.size_bytes(), json_opts, -1);
+              CUDF_EXPECTS(op.index >= 0, "Invalid numeric index specified in JSONPath");
+              op.expected_type = ARRAY;
+            }
+          }
+          return op;
+        }
+      } break;
+
+      // wildcard operator
+      case '*': {
+        pos++;
+        return path_operator{path_operator_type::CHILD_WILDCARD};
+      } break;
+
+      default: CUDF_FAIL("Unrecognized JSONPath operator"); break;
+    }
+    return {path_operator_type::ERROR};
+  }
+
+ private:
+  cudf::io::parse_options_view json_opts{',', '\n', '\"', '.'};
+
+  bool parse_path_name(string_view& name, string_view const& terminators)
+  {
+    switch (*pos) {
+      case '*':
+        name = string_view(pos, 1);
+        pos++;
+        break;
+
+      case '\'':
+        if (parse_string(name, false, '\'') != parse_result::SUCCESS) { return false; }
+        break;
+
+      default: {
+        size_t const chars_left = input_len - (pos - input);
+        char const* end         = std::find_first_of(
+          pos, pos + chars_left, terminators.data(), terminators.data() + terminators.size_bytes());
+        if (end) {
+          name = string_view(pos, end - pos);
+          pos  = end;
+        } else {
+          name = string_view(pos, chars_left);
+          pos  = input + input_len;
+        }
+        break;
+      }
+    }
+
+    // an empty name is not valid
+    CUDF_EXPECTS(name.size_bytes() > 0, "Invalid empty name in JSONpath query string");
+
+    return true;
+  }
+};
+
+/**
+ * @brief Preprocess the incoming JSONPath string on the host to generate a
+ * command buffer for use by the GPU.
+ *
+ * @param json_path The incoming json path
+ * @param stream Cuda stream to perform any gpu actions on
+ * @returns A pair containing the command buffer, and maximum stack depth required.
+ */
+std::pair<thrust::optional<rmm::device_uvector<path_operator>>, int> build_command_buffer(
+  cudf::string_scalar const& json_path, rmm::cuda_stream_view stream)
+{
+  std::string h_json_path = json_path.to_string(stream);
+  path_state p_state(h_json_path.data(), static_cast<size_type>(h_json_path.size()));
+
+  std::vector<path_operator> h_operators;
+
+  path_operator op;
+  int max_stack_depth = 1;
+  do {
+    op = p_state.get_next_operator();
+    if (op.type == path_operator_type::ERROR) {
+      CUDF_FAIL("Encountered invalid JSONPath input string");
```

Comment:
```suggestion
      CUDF_FAIL("Encountered invalid json_path input string");
```

Reply:
Technically, JSONPath is the name of the spec.  Not sure if it's important/useful to use it strictly like that though.  Your call.

https://tools.ietf.org/id/draft-goessner-dispatch-jsonpath-00.html

Reply:
That's fine. You used `json_path` in the API definition below. These probably should be consistent at least.